### PR TITLE
[workflow] workflow error handling 

### LINKED ISSUE #16990: error handling

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Implement error handling APIs and support for options.

Here options `step_max_retries` is for python application-level errors. For ray system error, we should use `max_retries`.

## Related issue number

Closes #16990 
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/step_function.py [hunk 1]
```diff
@@ -16,8 +16,15 @@
...
-    def __init__(self, func: Callable):
+    def __init__(self,
+                 func: Callable,
+                 step_max_retries=1,
+                 catch_exception=False,
+                 ray_options={}):
         self._func = func
+        self._step_max_retries = step_max_retries
+        self._catch_exception = catch_exception
+        self._ray_options = ray_options
```
Comment (suquark) [inline python/ray/experimental/workflow/step_function.py:27]: Checks if these inputs are valid (e.g. positive integers)

File: python/ray/experimental/workflow/step_function.py [hunk 2]
```diff
@@ -16,8 +16,15 @@
 class WorkflowStepFunction:
     """This class represents a workflow step."""
 
-    def __init__(self, func: Callable):
+    def __init__(self,
+                 func: Callable,
+                 step_max_retries=1,
+                 catch_exception=False,
+                 ray_options={}):
```
Comment (suquark) [inline python/ray/experimental/workflow/step_function.py:23]: if possible, we should not use non-basic objects as default value

**Review by suquark — APPROVED** [review]
LGTM

File: python/ray/experimental/workflow/tests/test_basic_workflows.py [hunk 3]
```diff
@@ -171,3 +172,64 @@ def test_run_or_resume_during_running(ray_start_regular_shared):
...
+    @workflow.step
+    def unstable_step():
+        v = int((tmp_path / "test").read_text())
+        (tmp_path / "test").write_text(f"{v + 1}")
+        if v < 10:
+            raise ValueError("Invalid")
+        return v
+
+    with pytest.raises(Exception):
+        unstable_step.options(step_max_retries=-1).step().run()
```
Comment (ericl) [inline python/ray/experimental/workflow/tests/test_basic_workflows.py:193]: How about just max_retries?

File: python/ray/experimental/workflow/tests/test_basic_workflows.py [hunk 4]
```diff
@@ -171,3 +172,64 @@ def test_run_or_resume_during_running(ray_start_regular_shared):
...
+    with pytest.raises(Exception):
+        unstable_step.options(step_max_retries=3).step().run()
+    assert 10 == unstable_step.options(step_max_retries=8).step().run()
+    (tmp_path / "test").write_text("0")
+    (ret, err) = unstable_step.options(
+        step_max_retries=3, catch_exception=True).step().run()
+    assert ret is None
+    assert isinstance(err, ValueError)
+    (ret, err) = unstable_step.options(
+        step_max_retries=8, catch_exception=True).step().run()
```
Comment (ericl) [inline python/ray/experimental/workflow/tests/test_basic_workflows.py:204]: catch_exceptions

File: python/ray/experimental/workflow/tests/test_basic_workflows.py (hunk 3 above)
Comment (iycheng) [reply python/ray/experimental/workflow/tests/test_basic_workflows.py:193]: I use step_max_retries to differentiate with ray internal errors on purpose. I gave some thought to this and I think it's good not to mix them.

Usually, when users specify the retry number here, they actually want to retry the application logic. Right now, max_retries has been used by ray, so I decide to use step_max_retries.

I'm thinking about hiding the ray internal retries from the user, but I'm not sure whether it's a good direction or not.

Right now, we don't have an easy way to detect ray internal errors without `ray.get` which I think is something missing and we probably want to add that later. It's an important feature that allows the workflow to detect errors in ray in a better way.

**Pull request merged**